providers/common/kdfs/kbkdf.c

```diff
@@ -0,0 +1,303 @@
+                                unsigned char *Context, size_t Context_len,
+                                unsigned char *K_i, size_t h, uint32_t L,
+                                unsigned char *KO, size_t KO_len)
+{
+    int ret = 0;
+    EVP_MAC_CTX *ctx = NULL;
+    size_t written = 0, to_write;
+    const unsigned char zero = 0;
+    uint32_t counter, i;
+
+    for (counter = 1; written < KO_len; counter++) {
+        i = htobe32(counter);
+
+        ctx = EVP_MAC_CTX_dup(ctx_init);
+        if (ctx == NULL)
+            goto done;
+
+        if (!EVP_MAC_update(ctx, (unsigned char *)&i, 4) ||
+            !EVP_MAC_update(ctx, Label, Label_len) ||
+            !EVP_MAC_update(ctx, &zero, 1) ||
+            !EVP_MAC_update(ctx, Context, Context_len) ||
+            !EVP_MAC_update(ctx, (unsigned char *)&L, 4))
+            goto done;
+
+        if (!EVP_MAC_final(ctx, K_i, NULL, h))
+            goto done;
+
+        to_write = KO_len - written;
+        memcpy(KO + written, K_i, MIN(to_write, h));
+        written += h;
+
+        EVP_MAC_CTX_free(ctx);
+        ctx = NULL;
+    }
+
+    ret = 1;
+done:
+    EVP_MAC_CTX_free(ctx);
+    return ret;
+}
+
+static int kbkdf_derive(void *vctx, unsigned char *key, size_t keylen)
+{
+    KBKDF *ctx = vctx;
+    OPENSSL_CTX *provctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+    int ret = 0;
+    const EVP_MD *md;
+    EVP_MAC *mac;
+    const char *mdname;
+    EVP_MAC_CTX *ctx_init;
+    OSSL_PARAM params[3];
+    unsigned char *K_i = NULL;
+    uint32_t L = htobe32(keylen * 8);
+    size_t h;
+
+    /* Label and Context are permitted to be empty. Check everything else. */
+    md = ossl_prov_digest_md(&ctx->digest);
+    if (md == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_MESSAGE_DIGEST);
+        return 0;
+    } else if (ctx->KI_len == 0 || ctx->KI == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    } 
+    mdname = EVP_MD_name(md);
+
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
+                                                 (char *)mdname, 0);
+    params[1] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY,
+                                                  ctx->KI, ctx->KI_len);
+    params[2] = OSSL_PARAM_construct_end();
+
+    mac = EVP_MAC_fetch(provctx, "HMAC", NULL);
+    if (mac == NULL)
+        return 0;
+    ctx_init = EVP_MAC_CTX_new(mac);
+    if (ctx_init == NULL)
+        goto done;
+
+    if (!EVP_MAC_CTX_set_params(ctx_init, params) || !EVP_MAC_init(ctx_init))
+        goto done;
+
+    h = EVP_MAC_size(ctx_init);
+    if (h == 0)
+        goto done;
+
+    K_i = OPENSSL_zalloc(h);
+    if (K_i == NULL)
+        goto done;
+
+    ret = kbkdf_derive_counter(
+        ctx_init, ctx->Label, ctx->Label_len, ctx->Context, ctx->Context_len,
+        K_i, h, L, key, keylen);
+done:
+    if (ret != 1)
+        OPENSSL_cleanse(key, keylen);
+    EVP_MAC_CTX_free(ctx_init);
```

Comment:
EVP_MAC_free(mac); will be required if you do a fetch

Reply:
Yes for this one.  It could be done earlier.